src/java.base/share/classes/java/util/HexFormat.java

```diff
@@ -0,0 +1,1023 @@
+/*
+ * Copyright (c) 2020, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package java.util;
+
+import jdk.internal.access.JavaLangAccess;
+import jdk.internal.access.SharedSecrets;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.CharBuffer;
+import java.nio.charset.CharacterCodingException;
+import java.nio.charset.StandardCharsets;
+
+/**
+ * Converts between bytes and chars and hex-encoded strings which may include additional
+ * formatting markup such as prefixes, suffixes, and delimiters.
+ * <p>
+ * There are two factories of {@code HexFormat} with preset parameters {@link #of()} and
+ * {@link #ofDelimiter(String) of(delimiter)}. For other parameter combinations
+ * the {@code withXXX} methods return copies of {@code HexFormat} modified
+ * {@link #withPrefix(String)}, {@link #withSuffix(String)}, {@link #withDelimiter(String)}
+ * or choice of {@link #withUpperCase()} or {@link #withLowerCase()} parameters.
+ * <p>
+ * For primitive to hexadecimal string conversions the {@code toHexDigits}
```

Comment:
It took me a little while to grok the naming convention being used here. It might be helpful to ensure that all the terms are clearly defined and used consistently throughout. 

"hexadecimal string"  - consists of just 0-9, a-f, A-F  (i.e. no delimiter, prefix, suffix)
"formatted hexadecimal string" - consists of hexadecimal string, and prefix, delimiter, suffix.

The `parseXXX` and `formatXXX` are logically related - and operate on "formatted hexadecimal strings"

The `toXXX` and `fromXXX` are logically related - one is kind of the inverse of the other. And these operate (even when accepting/producing chars) on "hexadecimal strings" - ignoring almost all properties of the HexFormat, except the toXXX cares about upper/lower case. Does `fromXXX` care about upper/lower ?

Reply:
Yes, naming is important and there are two dimensions in the api that are significant.  
Primitives vs byte arrays, and decoding and encoding.
```
         Primitives   ByteArrays
encode:   toXXX         format 
decode:   fromXXX       parse
```
Both fromXXX or parseXXX accept either upper or lower case.

The class javadoc and each method specifies whether it uses the prefix, suffix, delimiter, and upper/lower case choice.

The choice of 'formatted byte array' was used to make a strong distinction between formatting of byte arrays vs converting primitives with limited/no formatting.